src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -1026,7 +1027,9 @@ module Crystal
       block_visitor.path_lookup = path_lookup || current_type
       block_visitor.block_nest = @block_nest
 
+      @while_ensure_stack.push :block
```

Comment:
In `Crystal::CodeGenVisitor::Context` there is a pattern that might be useful here in `#with_context` the current value is kept in a local variable and then restored. That way we might don't event need an Array and leave the execution stack trace the push/pops in variables across the stack. There will be a need for a single ivar with the top of the stack for sure.

```crystal
def with_context(new_block_kind)
  old_block_kind = @block_kind
  @block_kind = new_block_kind
  value = yield
  @block_kind = old_block_kind
  value
end

def ...
  with_block_kind :block do
    node.body.accept block_visitor
  end
end
```